### PR TITLE
parser: maintain the supported character set in tidb

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/31761

at this moment, what charsets are supported is maintained by the parser, and cause some problems like https://github.com/pingcap/parser/issues/1140. 

it is not easy to use for the parser user, this pr implement the proposal https://internals.tidb.io/t/topic/456

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

```release-note
None
```

### CODE REVIEW

File: cmd/explaintest/r/new_character_set.result [hunk 1]
```diff
@@ -69,9 +69,9 @@ E4BDA0E5A5BD	7B22E6B58BE8AF95223A2022E4BDA0E5A5BD227D
 show charset;
 Charset	Description	Default collation	Maxlen
 ascii	US ASCII	ascii_bin	1
-binary	binary	binary	1
-gbk	Chinese Internal Code Specification	gbk_chinese_ci	2
-latin1	Latin1	latin1_bin	1
+binary	Binary pseudo charset	binary	1
```
Comment (Defined2014) [inline cmd/explaintest/r/new_character_set.result:72]: Seems need update mysql-test

File: util/collate/collate.go [hunk 2]
```diff
@@ -354,17 +355,17 @@ func CollationToProto(c string) int32 {
...
+}
+
+// All the names supported collations when new collation is not enabled should be in the following table.
+var supportedCollationWithoutNewCollation = []string{
+	charset.CollationUTF8,
+	charset.CollationUTF8MB4,
+	charset.CollationASCII,
+	charset.CollationLatin1,
+	charset.CollationBin,
+	charset.CollationGBKBin,
```
Comment (tangenta) [inline util/collate/collate.go:368]: Why not initializing the collations directly instead of names in `init()`?

Comment (xiongjiwei) [reply util/collate/collate.go:368]: we only need its name here

**Pull request closed without merging**